### PR TITLE
[RFC] New API: powerMonitor "shutdown" event

### PR DESCRIPTION
**Pull request by tarruda**
In this PR I add a new event to `powerMonitor`. Usage is very simple:

```js
powerMonitor.on('shutdown', (e) => {
  e.preventDefault()  // This will delay OS shutdown as much as possible
  someAppSpecificCleanup().then(() => {
    app.quit() // exit, this must be called ASAP!
  })
})
```

This is useful for any application that needs to ensure its state is saved correctly when the computer shuts down.

Currently only works on Linux, though I imagine windows/osx have similar APIs that could be used (I know I have seen Windows notification that some apps are delaying shutdown)

### CODE REVIEW

File: atom/browser/lib/power_observer_linux.cc [hunk 1]
```diff
@@ -100,6 +136,22 @@ void PowerObserverLinux::OnPrepareForSleep(dbus::Signal* signal) {
...
 }
 
+void PowerObserverLinux::OnPrepareForShutdown(dbus::Signal* signal) {
+  dbus::MessageReader reader(signal);
+  bool status;
+  if (!reader.PopBool(&status)) {
+    LOG(ERROR) << "Invalid signal: " << signal->ToString();
+    return;
+  }
+  if (status) {
```
Comment (alexeykuzmin) [inline atom/browser/lib/power_observer_linux.cc]: Sorry for nitpicking, but `status` is a really bad name for a boolean variable.
It's hard to guess that `true` or `false` values might mean.

File: lib/browser/api/power-monitor.js [hunk 2]
```diff
@@ -5,4 +5,16 @@ const {powerMonitor, PowerMonitor} = process.atomBinding('power_monitor')
 Object.setPrototypeOf(PowerMonitor.prototype, EventEmitter.prototype)
 EventEmitter.call(powerMonitor)
 
+powerMonitor.on('newListener', (event) => {
+  if (event === 'shutdown' && powerMonitor.listenerCount('shutdown') === 0) {
```
Comment (alexeykuzmin) [inline lib/browser/api/power-monitor.js]: Is the new "shutdown" listener added only _after_ this function is called?

File: spec/api-power-monitor-spec.js [hunk 3]
```diff
@@ -85,5 +85,34 @@ const skip = process.platform !== 'linux' || !process.env.DBUS_SYSTEM_BUS_ADDRES
         })
       })
     })
+
+    describe('when a listener is added to shutdown event', () => {
+      before(async () => {
+        const calls = await getCalls()
+        assert.equal(calls.length, 2)
+        powerMonitor.once('shutdown', () => { })
```
Comment (alexeykuzmin) [inline spec/api-power-monitor-spec.js]: Why all of it should be done in a `before` hook?
I don't see how it prepares an environment or context for tests.

File: lib/browser/api/power-monitor.js [hunk 4]
```diff
@@ -5,4 +5,16 @@ const {powerMonitor, PowerMonitor} = process.atomBinding('power_monitor')
...
+    powerMonitor.blockShutdown()
+  }
+})
+
+powerMonitor.on('removeListener', (event) => {
+  if (event === 'shutdown' && powerMonitor.listenerCount('shutdown') === 0) {
```
Comment (alexeykuzmin) [inline lib/browser/api/power-monitor.js]: Is a "shutdown" listener removed _before_ this function is called?

File: atom/browser/lib/power_observer_linux.cc (hunk 1 above)
Comment (tarruda) [reply atom/browser/lib/power_observer_linux.cc]: How about using `active` as shown in the [docs](https://www.freedesktop.org/wiki/Software/systemd/inhibit/)? I'm not sure if that name is better though.

File: lib/browser/api/power-monitor.js (hunk 2 above)
Comment (tarruda) [reply lib/browser/api/power-monitor.js]: Yes

File: lib/browser/api/power-monitor.js (hunk 4 above)
Comment (tarruda) [reply lib/browser/api/power-monitor.js]: Yes

File: spec/api-power-monitor-spec.js (hunk 3 above)
Comment (tarruda) [reply spec/api-power-monitor-spec.js]: Normally I try to structure BDD tests like this:

- `describe` describes the context
- `beforeEach` sets up the context
- `it` describes/verifies an expectation about the context

In this case the context is "when a listener is added to shutdown event" which is configured in the `before` callback.  The `it` call describes/verifies the "should call Inhibit to delay shutdown" expectation. I also added an assertion in the `before` callback to make it explicit that `getCalls` returns two calls prior to adding the listener.

Comment (alexeykuzmin) [reply spec/api-power-monitor-spec.js]: ok

File: atom/browser/lib/power_observer_linux.cc (hunk 1 above)
Comment (alexeykuzmin) [reply atom/browser/lib/power_observer_linux.cc]: `active` is better since it implies a binary state of active/inactive, while `status` can have numerous values, e.g. http status codes.

File: atom/browser/lib/power_observer.h [hunk 5]
```diff
@@ -18,7 +18,19 @@ namespace atom {
...
-typedef base::PowerObserver PowerObserver;
+class PowerObserver : public base::PowerObserver {
+ public:
+  PowerObserver() {}
+  void BlockShutdown() {}
+  void UnblockShutdown() {}
+  // Notification that the system is rebooting or shutting down. If the
+  // implementation returns true, the PowerObserver instance should try to delay
+  // OS shutdown so the application can perform cleanup before exiting.
+  virtual bool OnShutdown() { return false; }
```
Comment (ckerr) [inline atom/browser/lib/power_observer.h]: This code only works if the caller and the called are the same object (an instance of PowerMonitor inheriting from PowerObserver).. It's kind of a [template method](https://en.wikipedia.org/wiki/Template_method_pattern) via a crazy dependency between observer and monitor. Likewise, this is why Observer has new verb methods in this PR.

Mind you, I understand the choice -- we've already got the DBus code in Observer, so unless we refactor the DBus code into another place (eg a self-contained class or in PowerMonitorLinux) then Monitor has to have these Observer hooks.

This is going to be a pain to make future changes in and I'd love it if we could somehow clean up this dependency chain in this PR. That's a big ask though.

File: atom/browser/lib/power_observer_linux.cc (hunk 1 above)
Comment (tarruda) [reply atom/browser/lib/power_observer_linux.cc]: :+1:

File: atom/browser/lib/power_observer.h (hunk 5 above)
Comment (tarruda) [reply atom/browser/lib/power_observer.h]: I'm not sure I understand your request as I'm not very skilled in C++ design patterns. Can you give an example of what you have in mind?

File: atom/browser/lib/power_observer_linux.cc [hunk 6]
```diff
@@ -16,6 +16,10 @@ const char kLogindServiceName[] = "org.freedesktop.login1";
 const char kLogindObjectPath[] = "/org/freedesktop/login1";
 const char kLogindManagerInterface[] = "org.freedesktop.login1.Manager";
 
+// Store shutdown lock as a global, since we only want to release it when the
+// main process exits.
+base::ScopedFD shutdown_lock;
```
[the remark this answers is not in the document]
Comment (tarruda) [reply atom/browser/lib/power_observer_linux.cc]: I could move `shutdown_lock` into the class, but it is possible that the `ScopedFD` goes out of scope before the application has fully exited, which results in a small probability of systemd killing the app.

Maybe it is not a problem in practice since this class instance is probably deleted only a bit sooner than the app exits, but we would loose guarantee that the app exits cleanly by itself.

File: atom/browser/lib/power_observer_linux.cc [hunk 7]
```diff
@@ -54,12 +58,18 @@ void PowerObserverLinux::OnLoginServiceAvailable(bool service_available) {
...
+                           base::Bind(&PowerObserverLinux::OnPrepareForShutdown,
+                                      weak_ptr_factory_.GetWeakPtr()),
+                           base::Bind(&PowerObserverLinux::OnSignalConnected,
+                                      weak_ptr_factory_.GetWeakPtr()));
   logind_->ConnectToSignal(kLogindManagerInterface, "PrepareForSleep",
                            base::Bind(&PowerObserverLinux::OnPrepareForSleep,
                                       weak_ptr_factory_.GetWeakPtr()),
                            base::Bind(&PowerObserverLinux::OnSignalConnected,
                                       weak_ptr_factory_.GetWeakPtr()));
+  // Take sleep inhibit lock
```
[the remark this answers is not in the document]
Comment (tarruda) [reply atom/browser/lib/power_observer_linux.cc]: :+1:

File: spec/api-power-monitor-spec.js [hunk 8]
```diff
@@ -85,5 +85,34 @@ const skip = process.platform !== 'linux' || !process.env.DBUS_SYSTEM_BUS_ADDRES
...
+      })
+
+      describe('when PrepareForShutdown(true) signal is sent by logind', () => {
+        it('should emit "shutdown" event', (done) => {
+          powerMonitor.once('shutdown', () => { done() })
+          emitSignal('org.freedesktop.login1.Manager', 'PrepareForShutdown',
+            'b', [['b', true]])
+        })
+      })
+    })
```
[the remark this answers is not in the document]
Comment (tarruda) [reply spec/api-power-monitor-spec.js]: > What are the two calls that happen prior to adding the listener?

As I explained to @alexeykuzmin:

>  I also added an assertion in the before callback to make it explicit that getCalls returns two calls prior to adding the listener.

> If we keep the preventDefault approach, please add a test here to exercise it

I don't see how I could test this since there's no way of checking if a file descriptor was closed. (python-dbusmock doesn't even send a real file descriptor when it returns from an `Inhibit` request)

**Comment by tarruda** [comment]
> This seems like unnecessary hoops to jump through. I'd vote for removing this js bootstrap code and just auto-inhibiting in the C++ code, same as we do now for sleep.

The reason I enabled inhibiting only when a listener is added is because it would cause a small overhead when the OS shutdown even when no listener is added (since systemd has to wait until the ScopedFD is closed, which only happens when the app runs the JS event handler). I imagine there are many electron apps that don't care about cleanup code, it would be a shame if these apps caused unnecessary delays during shutdown.

I was planning on doing the same optimization for `suspend` event in a new PR after this one. In other words, `Inhibit` would only be called for `sleep` after a listener is added to `suspend` event.

I can leave as it is to reduce extra code, but IMHO it would not be good to make every electron app to add these inhibit locks unconditionally.

> PrepareForShutdown and PrepareForSleep have the same arguments and concepts, but we're handling one by releasing the lock immediately after emit and the other has a new custom handler. Is this to avoid breakage in how sleep is handled on macOS and Windows? 

Basically yes. If we allowed the app to prevent suspend with `e.preventDefault()`, we would need another API to allow the suspend to continue, and I'm not sure that would map well to macOS/Windows.

**Comment by tarruda** [comment]
> And, I know I've been a big pain already, but please document the API in docs/api/power-monitor.md. Thank you! :smile:

I was planning to write docs after finishing the code and addressing all reviews.

File: atom/browser/lib/power_observer_linux.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (tarruda) [reply atom/browser/lib/power_observer_linux.cc]: As I said in another comment, I planned to refactor the sleep Inhibit calls to use the same mechanism as shutdown Inhibit. So how about changing it to `BlockSleep`/`UnblockSleep` instead?

**Comment by tarruda** [comment]
I believe I addressed all comments, please let me know if something was missed

**Comment by zcbenz** [comment]
The API design matches macOS/Windows well, it would be straightforward to implement this API on macOS.

But the common implementation code does not fit other platforms' API designs, and most code can not be reused for other platforms, so I'm going to simplify the implementation by making certain code Linux.

**Comment by zcbenz** [comment]
I have added implementation for macOS.
